presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -315,17 +316,21 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
             rightPlanBuilder = rightPlanBuilder.appendProjections(rightComparisonExpressions, variableAllocator, idAllocator);
 
             for (int i = 0; i < leftComparisonExpressions.size(); i++) {
+                Expression leftExpression = leftPlanBuilder.rewrite(leftComparisonExpressions.get(i));
+                Expression rightExpression = rightPlanBuilder.rewrite(rightComparisonExpressions.get(i));
                 if (joinConditionComparisonOperators.get(i) == ComparisonExpression.Operator.EQUAL) {
                     VariableReferenceExpression leftVariable = leftPlanBuilder.translateToVariable(leftComparisonExpressions.get(i));
                     VariableReferenceExpression righVariable = rightPlanBuilder.translateToVariable(rightComparisonExpressions.get(i));
 
                     equiClauses.add(new JoinNode.EquiJoinClause(leftVariable, righVariable));
                 }
                 else {
-                    Expression leftExpression = leftPlanBuilder.rewrite(leftComparisonExpressions.get(i));
-                    Expression rightExpression = rightPlanBuilder.rewrite(rightComparisonExpressions.get(i));
                     postInnerJoinConditions.add(new ComparisonExpression(joinConditionComparisonOperators.get(i), leftExpression, rightExpression));
                 }
+                if (node.getType() == INNER) {
+                    postInnerJoinConditions.add(new IsNotNullPredicate(leftExpression));
+                    postInnerJoinConditions.add(new IsNotNullPredicate(rightExpression));
+                }
```

Comment:
I tried to add this optimisation in `PlanOptimizers` where all the other optimizers live . https://github.com/ssaumitra/presto/commit/07ca92e3f2355edbecedc40c06129302547787f8

The problem was that, `RelationPlanner` converts equality join expressions to `JoinNode.EquiJoinClause`. So `IsNotNullPredicate` can no longer be constructed over there. If there is a way to make that work, please let me know.

CC. @rschlussel

Reply:
Looks good.

When you are at it, also add it to the inner table of an outer join - so right side of left join and left side of right join can use the same optimization

Reply:
Makes sense. I will add that.